src/Catty/Functions/Text/RegularFunction.swift

```diff
@@ -0,0 +1,102 @@
+/**
+ *  Copyright (C) 2010-2020 The Catrobat Team
+ *  (http://developer.catrobat.org/credits)
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  An additional term exception under section 7 of the GNU Affero
+ *  General Public License, version 3, is available at
+ *  (http://developer.catrobat.org/license_additional_term)
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+class RegularFunction: DoubleParameterStringFunction {
+    static var tag = "REGEX"
+    static var name = "regex"
+    static var defaultValue = ""
+    static var requiredResource = ResourceType.noResources
+    static var isIdempotent = true
+    static let position = 85
+
+    func tag() -> String {
+        return type(of: self).tag
+    }
+
+    func firstParameter() -> FunctionParameter {
+        return .string(defaultValue: " an? ([^ .]+)")
+    }
+
+    func secondParameter() -> FunctionParameter {
+        return .string(defaultValue: "I am a panda")
+    }
+
+    func value(firstParameter: AnyObject?, secondParameter: AnyObject?) -> String {
+        let pattern = type(of: self).interpretParameter(parameter: firstParameter)
+        let longText = type(of: self).interpretParameter(parameter: secondParameter)
+
+        return self.regularExpression(pattern: pattern, longText: longText)
+    }
+
+    func formulaEditorSections() -> [FormulaEditorSection] {
+        return [.math(position: type(of: self).position)]
+    }
+
+    func regularExpression(pattern: String, longText: String) -> String {
+        var finalResult: String?
+        let regexOptions: NSRegularExpression.Options = []
+        do {
+            let regex = try NSRegularExpression(pattern: pattern, options: regexOptions)
+            let match = regex.firstMatch(in: longText, range: NSRange(0..<longText.utf16.count))
+            if match != nil {
+                if isParenthesesGroupPresent(in: pattern) {
```

Comment:
Thank you for your changes. Already looks much more readable. There is no need to use your custom _isParenthesesGroupPresent_ here. According to Apple's documentation of _firstMatch_, this method 

>  gives the overall matched range via its range property, and the range of each individual capture group via its range(at:) method.

So that means that this can be easily checked by using _match.numberOfRanges > 0_ since:

- match.range(at: 0) returns the full match
- match.range(at: 1) returns the first capture group